### PR TITLE
get brent's method to return `success`

### PR DESCRIPTION
**Pull request by smason**
This is a potential fix for issues #4630, just does the obvious thing.

I've not committed to scipy before, so let me know if you want tests that would fail for older versions…

### CODE REVIEW

**Comment by aarchiba** [comment]
Wouldn't hurt to have a couple of tests: one that checks that success sets the success attribute, and one that checks that failure doesn't (but also doesn't raise an exception). For the latter setting maxiter to a tiny number is probably necessary.

**Comment by smason** [comment]
OK, I've added a couple of tests.
Wasn't sure what exceptions were best to check for; "any" was easy so that's what I have done, but let me know if something else would be better!

**Comment by aarchiba** [comment]
Actually just raising an exception leads to a test failure, so I think there's no need to use a try/except block at all. I would prefer to see separate tests for separate features, so that if we break something we can see which aspects are broken (putting them all in one function means it stops with the first failing test).

**Comment by smason** [comment]
Yes, not sure why I tried to catch then throw in response, redundant as you say—will submit a patch that fixes that!
The tests I added were just adding things to a function which is documented as "combine all tests above for the minimize_scalar wrapper", it would seem messy to just have a couple of little tests separately for the same function.  Then again, there are quite a few tests there, maybe they should all be broken up?

File: scipy/optimize/tests/test_optimize.py [hunk 1]
```diff
@@ -721,6 +721,16 @@ def test_minimize_scalar(self):
         x = optimize.minimize_scalar(self.fun).x
         assert_allclose(x, self.solution, atol=1e-6)
 
+        x = optimize.minimize_scalar(self.fun, method='Brent')
+        assert(x.success)
```
Comment (dlax) [inline scipy/optimize/tests/test_optimize.py]: use `assert_` instead of `assert`

Comment (dlax) [inline scipy/optimize/tests/test_optimize.py]: Concerning separate test methods, this is fine as it is IMHO.

**Comment by smason** [comment]
Build seems be failing for reasons that aren't to do with my changes, is that normal?

**Comment by aarchiba** [comment]
I'm afraid that is normal; the testing gizmo gets randomly hung up sometime and the easiest fix is to just contribute more code. But I'd really prefer more granular tests. If there's nontrivial setup code it can go in a test fixture.

**Comment by dlax** [comment]
If one wants to split this big `test_minimize_scalar` method into small methods, that's certainly a good idea. Yet it seems a bit out of scope for this PR, which I think could be merged without this.

**Comment by smason** [comment]
I could submit another PR to split this up; there doesn't seem to be an obviously natural way to do this…
By optimisation method seems the most obvious, then another set to make sure the default method does the right thing.  Any preferences?

**Comment by aarchiba** [comment]
I guess yeah a new PR to break the method up is a good idea. I'd suggest using a test generator to run the same tests on each method (including the default) without duplicating code. It might even be practical to have the test generator generate the same tests for the non-wrapped functions.

**Comment by dlax** [comment]
Sam Mason wrote:
> I could submit another PR to split this up; there doesn't seem to be an
> obviously natural way to do this…
> By optimisation method seems the most obvious, then another set to make
> sure the default method does the right thing. Any preferences?

Splitting on empty line and setting test method name according to 
optimization method (plus possible details) would be fine IMO.

**Comment by smason** [comment]
Denis Laxalde wrote:

> Splitting on empty line and setting test method name according to
> optimization method (plus possible details) would be fine IMO.


That would end up in a *lot* (~15) of little tests, is that really right?

**Comment by smason** [comment]
Anne Archibald wrote:
> I'd suggest
> using a test generator to run the same tests on each method (including the
> default) without duplicating code.

Doing this for basic features seems reasonable.  I don't know all the
methods too well yet, but I presume some will have specific options
and hence will still need a few tests dedicated to each method.

> It might even be practical to have the
> test generator generate the same tests for the non-wrapped functions.

Would the point of this be to clean up the existing test code?  Seems
like they are being tested already.

**Comment by dlax** [comment]
Sam Mason wrote:
> That would end up in a *lot* (~15) of little tests, is that really right?

Yes.

**Comment by aarchiba** [comment]
I'd say this PR is good to go. When it's done I'll see if I can clean up the testing. As you say, this isn't particularly a matter of adding new tests so much as making sure it's easy to tell exactly which bit breaks when tests start failing.

**Comment by smason** [comment]
I'd be happy to split the tests up if a consensus can be reached about how best to organise them.  The generator approach seems sensible for testing each method systematically, but I don't know SciPy well enough to comment on much else!

**Comment by dlax** [comment]
The generator approach does not really seem obvious to me here. It's not like we could easily do something like:
```python
def check(param):
  [compute using minimize_scalar]
  [assert]

for p in parameters:
  yield check, p
```
since almost all `minimize_scalar` calls are different.
And doing something more complicate with nose generators is problably not a very good idea since this will eventually get pretty hard to debug if something breaks (see section "How do I write tests with test generators?" in http://docs.scipy.org/doc/scipy-dev/reference/hacking.html)

**Comment by aarchiba** [comment]
I was picturing:
```python
def test_minimize_scalars():
    for method in [None, 'brent', 'golden', 'bounded']:
       yield check_minimization_works, method
       yield check_failure_returned, method

def check_minimization_works(method):
    [compute using minimize_scalar, method, the standard function, bracket, and result]
    [assert success is reported]
    [assert the return value is close]
```
And then tests for the features of specific methods not covered by these (e.g. what happens with out-of-bounds values for bounded minimization).

**Comment by smason** [comment]
@aarchiba, yes that's the sort of thing I would have expected—should clean up the current tests a bit.

**Pull request merged**